aws_protocol_adaptor/device_client/aws_nvmsgbroker.c

```diff
@@ -243,41 +243,32 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
 	return NVDS_MSGAPI_OK;
 }
 
-/* ************************************************************************* */
-// Subscribe def
-/* ************************************************************************* */
-
-void subscribe_callback_handler(AWS_IoT_Client *pClient, char *topicName, uint16_t topicNameLen,
-								IoT_Publish_Message_Params *params, void *pData)
-{
-	IOT_INFO("Subscribe callback");
-	IOT_INFO("%.*s\t%.*s", topicNameLen, topicName, (int)params->payloadLen, (char *)params->payload);
-}
-
 NvDsMsgApiErrorType nvds_msgapi_subscribe(NvDsMsgApiHandle h_ptr, char **topics, int num_topics, nvds_msgapi_subscribe_request_cb_t cb, void *user_ctx)
 {
 	IOT_INFO("Subscribe called\n");
 
 	if ((h_ptr == NULL) || (topics == NULL) || (num_topics <= 0))
 	{
-		IOT_ERROR("Essensial args missing for function nvds_msgapi_subscribe: %d, %d, %d\n", (h_ptr == NULL), (topics == NULL), (num_topics == NULL);
+		IOT_ERROR("Essential args missing for function nvds_msgapi_subscribe: %d, %d, %d\n", (h_ptr == NULL), (topics == NULL), (num_topics == NULL);
 		return NVDS_MSGAPI_ERR;
 	}
 	if (!cb)
 	{
-		IOT_ERROR("Callback function for nvds_msgapi_send cannot be NULL\n");
+		IOT_ERROR("Callback function for nvds_msgapi_subscribe cannot be NULL\n");
 		return NVDS_MSGAPI_ERR;
 	}
 	IoT_Error_t rc = FAILURE;
 	AWS_IoT_Client *client = (AWS_IoT_Client *)h_ptr;
-	rc = aws_iot_mqtt_subscribe(client, topics, strlen(topics), QOS0, subscribe_callback_handler, NULL);
-	if (SUCCESS != rc)
+	for (int i = 0; i < num_topics; i++)
 	{
-		IOT_ERROR("Unable to subscribe, error: %d\n", rc);
-		return NVDS_MSGAPI_ERR;
+		rc = aws_iot_mqtt_subscribe(client, topics[i], strlen(topics[i]), QOS0, cb, user_ctx);
```

Comment:
This will not compile. `cb` is not the correct type. You will likely need to somehow wrap `cb` in a function that implements the correct type expected by `aws_iot_mqtt_subscribe`.

Recall that this is the callback type accepted by `aws_iot_mqtt_subscribe`:

```c
typedef void(* pApplicationHandler_t) (AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
```

And this is the callback type accepted by `nvds_msgapi_subscribe`:

```c
typedef void(* nvds_msgapi_subscribe_request_cb_t)(NvDsMsgApiErrorType flag, void *msg, int msg_len, char *topic, void *user_ptr)
```

So we'll need to create a function of type `*pApplicationHandler_t` that calls `cb`.

Something **like** this (not this exactly) would work:

```c
void nvds_msgapi_subscribe_cb_wrapped(AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
{
    cb(NVDS_MSGAPI_OK, pParams->payload, pParams->payloadLen, pTopicName, pClientData);
}
```

The trick here is determining how to create this wrapped callback. You will likely need to have a static variable that stores a pointer to such a function, then you can set the value of this static variable when the subscribe method is called for the first time.

Reply:
@Shubhamturakhia Here's an idea:

Define a static variable `nvds_cb` that we will use to store the callback provided by a user:

```c
static nvds_msgapi_subscribe_request_cb_t nvds_cb;
```

Define a function that implements the type of callbacks expected by the mqtt subscribe function:

```c
void nvds_cb_wrapped(AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
{
    nvds_cb(NVDS_MSGAPI_OK, pParams->payload, pParams->payloadLen, pTopicName, pClientData);
}
```

In your implementation of nvds_msgapi_subscribe, assign the callback supplied as an argument to the `nvds_cb` variable:

```c
nvds_cb = cb;
```